### PR TITLE
[GRPO] create a new resharding API which uses split_my_mesh_axis from pathwaysutils

### PR DESCRIPTION
**Pull request by khatwanimohit**
# Description

This PR introduces an API for resharding which uses split_by_mesh_axis from pathwaysutils. It includes a function to find intermediate shardings for split_by_mesh_axis when possible. 

Logs for llama3.1-8b on v5p-64:
https://cloudlogging.app.goo.gl/cu9xgZtatFSUyfyh6

If the change fixes a bug or a Github issue, please include a link, e.g.,:
FIXES: b/449217704

*Notice 1:* Once all tests pass, the "pull ready" label will automatically be assigned.
This label is used for administrative purposes. Please do not add it manually.

*Notice 2:* For external contributions, our settings currently require an approval from a MaxText maintainer to trigger CI tests.

# Tests

A test which runs on v5p-64 with pathways as backend

# Checklist

Before submitting this PR, please make sure (put X in square brackets):
- [x] I have performed a self-review of my code. For an optional AI review, add the `gemini-review` label.
- [x] I have necessary comments in my code, particularly in hard-to-understand areas.
- [x] I have run end-to-end tests tests and provided workload links above if applicable.
- [x] I have made or will make corresponding changes to the doc if needed.

### CODE REVIEW

File: src/MaxText/experimental/rl/grpo_trainer.py [hunk 1]
```diff
@@ -738,21 +715,40 @@ def generation_worker_fn(
...
       try:
-        with jax.profiler.StepTraceAnnotation("inference"):
-          generate_completions(
+        if worker_step == profiler_object.start_initial_profile_step and not is_profiling:
+          profiler_object.activate()
+          is_profiling = True
+        elif worker_step == profiler_object.finished_initial_profile_step and is_profiling:
+          profiler_object.deactivate()
+          is_profiling = False
+        with jax.profiler.StepTraceAnnotation("inference", step_num=worker_step):
```
Comment (xuefgu) [inline src/MaxText/experimental/rl/grpo_trainer.py:747]: Maybe remove since it's not working as intended?

On a general note, should we treat profiling in a separate PR and focus on functionality in this one?

File: src/MaxText/experimental/rl/grpo_trainer.py [hunk 2]
```diff
@@ -830,7 +814,7 @@ def generation_worker_fn(
               {"params": state.params["params"]},
               {"params": state_mesh_shardings.params["params"]},
               mesh,
-              inference_state_mesh_shardings,
+              {"params": inference_state_mesh_shardings.params["params"]},
```
Comment (xuefgu) [inline src/MaxText/experimental/rl/grpo_trainer.py:834]: Should we clear the rollout buffer after the reshard call?

File: src/MaxText/experimental/rl/grpo_trainer.py (hunk 1 above)
Comment (khatwanimohit) [reply src/MaxText/experimental/rl/grpo_trainer.py:747]: I'd still like to keep this call since it doesn't hurt and it shows that call on the cpu-trace.

File: src/MaxText/experimental/rl/grpo_trainer.py (hunk 2 above)
Comment (khatwanimohit) [reply src/MaxText/experimental/rl/grpo_trainer.py:834]: done

File: end_to_end/tpu/test_grpo.sh [hunk 3]
```diff
@@ -23,10 +27,11 @@ JAX_BACKEND_TARGET=grpc://127.0.0.1:29000
 ENABLE_PATHWAYS_PERSISTENCE='1'
 HF_TOKEN=${HF_TOKEN}
 
-MAX_PREFILL_LENGTH=128
-MAX_TARGET_LENGTH=256
+MAX_PREFILL_LENGTH=${MAX_PREFILL_LENGTH:-128}
+MAX_TARGET_LENGTH=${MAX_TARGET_LENGTH:-256}
```
Comment (A9isha) [inline end_to_end/tpu/test_grpo.sh:31]: I don't quite understand what these two lines are doing

File: end_to_end/tpu/test_grpo.sh [hunk 4]
```diff
@@ -23,10 +27,11 @@ JAX_BACKEND_TARGET=grpc://127.0.0.1:29000
...
 NUM_GENERATIONS=2
 
+INFERENCE_PER_DEVICE_BS=$((${INFERENCE_PER_DEVICE_BATCH_SIZE} * ${NUM_GENERATIONS}))
```
Comment (A9isha) [inline end_to_end/tpu/test_grpo.sh:34]: nit: maybe remove the parentheses

File: end_to_end/tpu/test_grpo.sh (hunk 3 above)
Comment (khatwanimohit) [reply end_to_end/tpu/test_grpo.sh:31]: these lines are setting default values of prefill and target length to 128 and 256 if the user does not provide them as an argument to the bash script

File: end_to_end/tpu/test_grpo.sh (hunk 4 above)
Comment (khatwanimohit) [reply end_to_end/tpu/test_grpo.sh:34]: I think this is a bash syntax if I want to perform the multiplication operation using bash's env variables.

Comment (A9isha) [reply end_to_end/tpu/test_grpo.sh:34]: ah I see - thank you

File: end_to_end/tpu/test_grpo.sh (hunk 3 above)
Comment (A9isha) [reply end_to_end/tpu/test_grpo.sh:31]: ah I see - thank you

File: src/MaxText/experimental/rl/grpo_trainer.py [hunk 5]
```diff
@@ -586,36 +588,17 @@ def generate_completions(
...
-  a batch of prompts from the data loader, uses the offline inference engine to
-  generate multiple completions for each prompt, processes the results, and adds
-  them to a shared data buffer for the training loop to consume.
-
-  Args:
-    worker_data_loader: The DataLoader instance for fetching prompt data.
-    worker_inference_engine: The offline inference engine for generation.
-    worker_tokenizer_model: The tokenizer model.
-    worker_config_inference: The configuration for the inference process.
-    worker_config_train: The main training configuration.
```
Comment (A9isha) [inline src/MaxText/experimental/rl/grpo_trainer.py:606]: why are we removing the arg descriptions for all the args?

File: src/MaxText/experimental/rl/grpo_trainer.py [hunk 6]
```diff
@@ -738,21 +717,40 @@ def generation_worker_fn(
       worker_input_data_shardings: Sharding specs for the generated data.
       engine_lock: A lock for thread-safe inference engine access.
       stop_event: A threading.Event to signal when the worker should stop.
+      profiling_event: a threading.Event to signal when to profile.
```
Comment (A9isha) [inline src/MaxText/experimental/rl/grpo_trainer.py:735]: is this referring to `profiler_object`?

File: src/MaxText/experimental/rl/grpo_trainer.py (hunk 5 above)
Comment (khatwanimohit) [reply src/MaxText/experimental/rl/grpo_trainer.py:606]: added them back

File: src/MaxText/experimental/rl/grpo_trainer.py (hunk 6 above)
Comment (khatwanimohit) [reply src/MaxText/experimental/rl/grpo_trainer.py:735]: this is referring to threading event because profiling is happening inside a separate thread other than the main thread

**Review by A9isha — APPROVED** [review]
